src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -2,30 +2,48 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
 
-namespace System
+namespace System.Net
 {
     internal static partial class IPv4AddressHelper
     {
         internal const long Invalid = -1;
         private const long MaxIPv4Value = uint.MaxValue; // the native parser cannot handle MaxIPv4Value, only MaxIPv4Value - 1
+
         private const int Octal = 8;
         private const int Decimal = 10;
         private const int Hex = 16;
 
         private const int NumberOfLabels = 4;
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static ushort ToUShort<TChar>(TChar value)
+            where TChar : unmanaged, IBinaryInteger<TChar>
+        {
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
+            return typeof(TChar) == typeof(char)
+                        ? (char)(object)value
+                        : (byte)(object)value;
+        }
+
         // Only called from the IPv6Helper, only parse the canonical format
-        internal static int ParseHostNumber(ReadOnlySpan<char> str, int start, int end)
+        internal static int ParseHostNumber<TChar>(ReadOnlySpan<TChar> str, int start, int end)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
             Span<byte> numbers = stackalloc byte[NumberOfLabels];
 
             for (int i = 0; i < numbers.Length; ++i)
             {
                 int b = 0;
-                char ch;
+                ushort ch;
```

Comment:
Thanks for this. I've been testing with SharpLab a little though, and I've not yet been able to do anything to eliminate these zero-extensions - [link](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdAHICuYApgE5QDGKA3EaRdQEp0YCKEyoBhAPZgADlAA2LAMosAbu0Zce5JQAsIzGQBkIwKoOGjG3QjwDMFbGXFkA3kTKeyAbQCyjBF1JABMASVl5AAp/QJDwmXkAeRkRSRgUKgBBAHNs5k0UKFVGUJh5WFhsgEoAXQ8vbwApKAQAcUYYFnZIhABPGUZJADNItn1mKqq4Mma2jq62Hv7BkeBehEZJusIvCgc6FCDmBGc04uOAHgAVcXGAPkib8bI2KvrdzwB3XRZGMieDGQQGQhJAYBBsoxgtNQgAhWAGXqlDaQ5jXW4GO7vD4AXjuZD6A2GjwxEzIOJxBOWxLGBjeOw+jIA/GRRuMqpFJMAAFaMNgIKpsbGMzzAyJrDYcrm8/mCmy7bFNFrtTqsRaElZsulTGbK+ZqpZE1brTa1bGYByYAAsZBVCyUUGynTCwkYqPR90i/EYEGCiTKvSUMggMA9mJe4xQ9JFZG+v3+pKBIJgYIhUJh8PBzCRrvdAOYWIZnncRY+Q0kzFZsBOUHJZBInDItYuEYMGSMHWygUbUAA1L3oyKSzHdtWXmw61IYOcEFrmChvFAalV5SPPAAxZjSUmjV6rxkAX2xR9LitmdoNGpp7J15/13SvxslZtLFooNov7AAqulHc7kW6LBhgWXo+n6AZBiGwH4rS86Doycb5AmzzAqCIZptCZBwgi2YAXmpKFrsw4iuWlaRGOtaUg2TZkC2sHtp23ZNv28EfMRa4HEcJxsBOlJTjOc4LkuK7CiKm7buMu4iaWuwnrJ2KKjEQRhBE0QBMp8RJCkUBpBkOR5AURQlGUFQwNU2y7G+lAAGxNsIZDiWAO5jq82LsR8mAAOwvGQdx8eMVySAAEowAAeRiSJIADWdAyFQHZmYEZAsiQoXruuSakoFIXhZFMUyN4bA1PuZByV45qWrgtner6/ryIGwahhKjD4llwVhRF0WxeShENKJnipel0yDeuw1paN9bjWNQ2TTNI3TRN82zYtU3LQt61rZtAD0W1kLgACs/WbUtJ2radc1nZdF3XStN0bedE07WQdgEDJXgPfdV23d9n13cdX2/T9ZBPVanlHalJDTJNuBQ6l2Cw6FdgI1aCP7Qj1kI55CMABwIwAnAjf0ff9M1PdZdjg5dmQI7CCPiAjAAiCMAKKE4TQPE5zAPbbtnl45TRPc1zgsixzQvc09eOHW9A2XRACPAAjbAI5hk2MAjQzs4D2sk2LpO7bghsC3rJs68Lpu6zrT24NgYMy5bDvm2b4ui1bBtWhT9tO47LsW973vW/t/Ne77zuuz74cBwbnnSzG/uhxHfsJ8T1t469cfJ5nkdZ0n+tkNgJB2xn2cl7nZdh49u3YNgnvF+XicV43DfA1XdjB3XTfx6XndfZ4T3YPtsdeMVRAHkAA=). The MOVZX here is taking place because of the width of the `TChar` type, not that of the intermediate variable, so swapping the data type wouldn't do anything on this platform.

Are zero-extensions eliminated under other circumstances - and if they aren't, should I still make these type changes?

Reply:
What I had in mind are [cases like this](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YA5Dvj6CwuabUbMWAJQ7CMg5SwDCEfAAdBAGz4BlPgDdVMdTMZ2AFmIsAZbMC6+obKGjIAzEykDCY0AN40DEkMANoAsjAYnhAAJgCSltYAFBlZuQUW1gDyFoYQwgIAggDmzbC4uIKOMHnC1iIizQCUALqJySIYfKLWckgMHLjZUBgMACoQAKpe0BgAPGsm3lAAfEWHxwyO2NYcMEPjyckA7p58MOtHYgwgC8L42FEzRgOTQDDybBEYnEvSmwKgBy+p0eDAS1CeT2IAHYGBhxBYYBAuOckUMGABecm4/GE4lgY4PdEY5kAfgYRXpYiGRQgwAAVjAwBghtdbjAUcykr8isBxFNubyBUKRTc7hongBfFEolIAKUEGAA4jBhCowEU8QSiTK5fdRijJtMbnMGJMGAAFMS4GAACQguAwinwwD4iOOZx0MGwOSqfXEdgsgLDYhODADUDBboDYgwmYMDBNOUZkoYr3en0uv30AKBILBEKhElhMHhyeRTKSaJLCaTsqmqeEShD/ApaYw2DAAGsbtYIGAGH2YCkUCMwh2MVxoOy3YJR3RJK6GHsGIPg3wBL4Tc0sgeANS3wTFyVdksTfPAPfq19JRbLVb0tdvySTcoHZA8imzFYjwLYQiwYAAyeD2XpUcNm2P8IIwKAUkgjARiGMkAEIqQAchYEiEKQlDiIYEiQBIu9b1wp9vxfIDkg/KkZQYAAqBhyDoMlbwYFC4FougGIlSUtWoKTmVPYdcBSQQRlHG15WAL9v3vXCtIxGS5KYHFIVECR3SgIwDS6VxdCjfIDHCUhIWaABRWDBEBIoFPPIY9IM9c9QNY1TQs81LVpdS7TGddHSgGYXTdT1+F9f1AyHPhSDbCM7Njax40TYQsrHDNXXzXC81WQsWIxMtYArb4q3+QFsGBUFwRM6Fm1bC4UxRNiMR7QrFwHdKRypbMpxnOcF1tZdVxoQyQO3fNdypfdD2Pbz+BYS9hGvTxGMfQz+pLN1OIYfdDKeN0AIW9cSyWopwNw6Cqso5DPFQrYdhWTDsNw/CiNI8j3uo0j6MY5iruSE6gPO7i+IEoSRM+sSSIkvSS389itqUlS1MXIZNOhpIdPHFZMeSbHJWxBgOrMiz8Cs7oBEjaNYUc5y3JyDzhC80bcF8lEZI1IA==).

Reply:
I see what you mean! Thanks, I'll make the changes and revalidate my benchmark,

Reply:
Benchmark results are pretty promising - I've updated my earlier reply. The ParseHostNumber method you've pointed out shows the largest improvement, but there's a ~2ns pattern of improvement across the board.